### PR TITLE
Awesome IPLD endeavour

### PR DESCRIPTION
**Pull request by diasdavid**
This makes js-ipfs understand how to use IPLD.

As a compromise, had to turn off bitswap as its migration to IPLD is not trivial. We have a focused milestone for it (Bitswap complete).

The last thing to add to this PR is the new `dag` API.

### CODE REVIEW

File: src/core/components/block.js [hunk 1]
```diff
@@ -2,48 +2,61 @@
...
       })
     }
   }
 }
 
-function cleanHash (hash) {
-  if (typeof hash === 'string') {
-    return multihash.fromB58String(hash)
+function cleanCid (cid) {
+  if (cid.constructor.name === 'CID') {
```
Comment (dignifiedquire) [inline src/core/components/block.js]: please use `require('cids').isCID` instead

**Comment by diasdavid** [comment]
PR Table from: https://github.com/ipld/js-ipld-resolver/pull/60#issuecomment-255381311

PR                                                                                      | CR | RFM | M
------------------------------------------------------|----|------|---- 
https://github.com/ipld/js-ipld-resolver/pull/60            | ✔️  | ✔️     | ✔️
https://github.com/ipld/js-ipld-dag-pb/pull/1                | ✔️  | ✔️     | ✔️
https://github.com/ipld/js-ipld-dag-cbor/pull/25          | ✔️  | ✔️     | ✔️ 
https://github.com/ipfs/js-cid/pull/4                              | ✔️  | ✔️      | ✔️
https://github.com/ipfs/js-ipfs-unixfs-engine/pull/74  | ✔️  | ✔️      |
https://github.com/multiformats/multicodec/pull/16    |       |          |
https://github.com/ipfs/js-ipfs-block/pull/5                  | ✔️   | ✔️     | ✔️
https://github.com/ipfs/js-ipfs-block-service/pull/27  | ✔️   | ✔️     | ✔️
https://github.com/ipfs/js-ipfs-repo/pull/89                 | ✔️   | ✔️     | ✔️
https://github.com/ipfs/interface-ipfs-core/pull/78     | ✔️   | ✔️     |        
https://github.com/ipfs/js-ipfs/pull/532                        | ✔️   | ✔️     |
https://github.com/ipfs/js-ipfs-api/pull/398                 | ✔️   | ✔️     |

**Comment by diasdavid** [comment]
Can I get some CR ? :)

**Comment by diasdavid** [comment]
Also added as a bonus, a refreshed readme :)

**Comment by dignifiedquire** [comment]
@diasdavid why the red CI?

**Comment by diasdavid** [comment]
@dignifiedquire it is that 'tmpdir' thing, I was grepping through to find where it was coming from, but with no luck

File: src/cli/commands/block/put.js [hunk 2]
```diff
@@ -22,7 +22,7 @@ function addBlock (buf) {
         throw err
       }
 
-      console.log(bs58.encode(block.key).toString())
+      console.log(bs58.encode(block.key()).toString())
```
Comment (dignifiedquire) [inline src/cli/commands/block/put.js]: Why doesn't this throw, shouldn't `block.key()` be async?

File: src/cli/commands/object/patch/add-link.js [hunk 3]
```diff
@@ -15,23 +15,41 @@ module.exports = {
...
+          }
+          linkedObj.multihash((err, multihash) => {
+            if (err) {
+              throw err
+            }
+
+            const link = new DAGLink(argv.name, size, multihash)
+
+            ipfs.object.patch.addLink(argv.root, link, {enc: 'base58'})
+            .then((node) => {
```
Comment (dignifiedquire) [inline src/cli/commands/object/patch/add-link.js]: let's not use promises mixed with callback

File: test/cli/test-bitswap.js [hunk 4]
```diff
@@ -10,7 +10,7 @@ const createTempNode = require('../utils/temp-node')
 const repoPath = require('./index').repoPath
 const ipfs = require('../utils/ipfs-exec')(repoPath)
 
-describe('bitswap', () => {
+describe.skip('bitswap', () => {
```
Comment (dignifiedquire) [inline test/cli/test-bitswap.js]: 😿

**Pull request merged**